component/sqs/component.go

```diff
@@ -0,0 +1,332 @@
+// Package sqs provides a native consumer for AWS SQS.
+package sqs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
+	"github.com/beatlabs/patron/correlation"
+	"github.com/beatlabs/patron/log"
+	"github.com/beatlabs/patron/trace"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ProcessorFunc definition of a async processor.
+type ProcessorFunc func(context.Context, Batch)
+
+type messageState string
+
+const (
+	sqsAttributeApproximateNumberOfMessages           = "ApproximateNumberOfMessages"
+	sqsAttributeApproximateNumberOfMessagesDelayed    = "ApproximateNumberOfMessagesDelayed"
+	sqsAttributeApproximateNumberOfMessagesNotVisible = "ApproximateNumberOfMessagesNotVisible"
+	sqsAttributeSentTimestamp                         = "SentTimestamp"
+
+	sqsMessageAttributeAll = "All"
+
+	consumerComponent = "sqs-consumer"
+
+	ackMessageState     messageState = "ACK"
+	nackMessageState    messageState = "NACK"
+	fetchedMessageState messageState = "FETCHED"
+)
+
+var (
+	messageAge     *prometheus.GaugeVec
+	messageCounter *prometheus.CounterVec
+	queueSize      *prometheus.GaugeVec
+)
+
+func init() {
+	messageAge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_age",
+			Help:      "Message age based on the SentTimestamp SQS attribute",
+		},
+		[]string{"sqsAPI"},
+	)
+	prometheus.MustRegister(messageAge)
+	messageCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_counter",
+			Help:      "Message counter",
+		},
+		[]string{"sqsAPI", "state", "hasError"},
+	)
+	prometheus.MustRegister(messageCounter)
+	queueSize = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "queue_size",
+			Help:      "Queue size reported by AWS",
+		},
+		[]string{"state"},
+	)
+	prometheus.MustRegister(queueSize)
+}
+
+// Component implementation of a async component.
+type Component struct {
```

Comment:
would it make sense to split these attributes based on usage ? Might be easier to read.
i.e. (not the best naming, just to visualise the point)
```
type Component struct {
     queueInfo queueInfo
     stats stats
     config config
     failureStrategy failureStrategy
}

type queueInfo struct {
    name string
    url string
 ....
}

type stats struct {
    interval time.Duration
 ....
}

```

Reply:
done